### ISSUE TITLE
Add conflicts-with field to .cabal file

### ISSUE THREAD

**Issue by tibbe** [haskell/cabal#2064]
A `conflicts-with` flag would allow you to express that e.g. two packages are incompatible because they export the same module (e.g. network-2.5 and network-uri-2.6).

**Comment by 23Skidoo** [comment]
It'd be nice to make the solver use this information, and while it can be inferred from the package index, it probably makes sense to cache it. Right place to do it would be during `cabal update`.

**Comment by 23Skidoo** [comment]
One complication here is that the list of exported modules can be affected by flag choice. So we need to record that info as well. The cached conflicts info file then could look like this:

```
foo-1.0: bar-1.0
bar-1.0: foo-1.0
baz-1.0 +abc: quux-1.0, quux-2.0
[...]
```

**Comment by hvr** [comment]
See also #3061 which discusses alternatives to this "negative" `conflicts-with` form. 

Btw, I think the motivating use-case of *generally* conflicting modules is not a good idea at all. In fact, I've got a couple of cases where I deliberately depend on packages with conflicting module names which would stop working if cabal started automatically outlawing such cases. Conflicting modules names are no different than conflicting symbols and we have got facilities to disambiguate. In the case of network-uri/network it's more of a convenience to encode negative constraints, and hence it's a special corner case.
